File: morphir-dotnet/src/Morphir.CodeModel/IR/AccessControlled.fs
/// Module to manage access to a node in the IR. This is only used to declare access levels
/// not to enforce them. Enforcement can be done through the helper functions
/// [withPublicAccess](#withPublicAccess) and [withPrivateAccess](#withPrivateAccess) but it's
/// up to the consumer of the API to call the right function.
module Morphir.IR.AccessControlled

open Morphir.SDK.Maybe

/// <summary>
/// Type that represents different access levels.
/// </summary>
type AccessControlled<'a> = { Access: Access; Value: 'a }

/// <summary>
/// Public or private access.
/// </summary>
and Access =
    | Public
    | Private

let accessControlled access value = { Access = access; Value = value }

/// Mark a node as public access. Actors with both public and private access are allowed to see.
let inline ``public`` value = { Access = Public; Value = value }

/// Mark a node as private access. Only actors with private access level can see.
let inline ``private`` value = { Access = Private; Value = value }

/// Mark a node as public access. Actors with both public and private access are allowed to see.
let inline mkPublic value = { Access = Public; Value = value }

/// Mark a node as private access. Only actors with private access level can see.
let inline mkPrivate value = { Access = Private; Value = value }

///  Get the value with public access level. Will return `Nothing` if the value is private.
let withPublicAccess ac =
    match ac.Access with
    | Public -> Just ac.Value
    | Private -> Nothing

/// Get the value with private access level. Will always return the value.
let withPrivateAccess ac =
    match ac.Access with
    | Public -> ac.Value
    | Private -> ac.Value

// Get the value with public or private access level. Will return `Nothing` if the value is private
// and accessed using public access.
let withAccess access ac =
    match access with
    | Public -> withPublicAccess ac
    | Private -> Just(withPrivateAccess ac)

let map (f: 'a -> 'b) (ac: AccessControlled<'a>) : AccessControlled<'b> = {
    Access = ac.Access
    Value = f (ac.Value)
}

File: morphir/ir/core/src/morphir/ir/AccessControlled.scala
package morphir.ir

/** Generated based on IR.AccessControlled
*/
object AccessControlled{

  sealed trait Access {
  
    
  
  }
  
  object Access{
  
    case object Private extends morphir.ir.AccessControlled.Access{}
    
    case object Public extends morphir.ir.AccessControlled.Access{}
  
  }
  
  val Private: morphir.ir.AccessControlled.Access.Private.type  = morphir.ir.AccessControlled.Access.Private
  
  val Public: morphir.ir.AccessControlled.Access.Public.type  = morphir.ir.AccessControlled.Access.Public
  
  final case class AccessControlled[A](
    access: morphir.ir.AccessControlled.Access,
    value: A
  ){}
  
  def map[A, B](
    f: A => B
  )(
    ac: morphir.ir.AccessControlled.AccessControlled[A]
  ): morphir.ir.AccessControlled.AccessControlled[B] =
    (morphir.ir.AccessControlled.AccessControlled(
      ac.access,
      f(ac.value)
    ) : morphir.ir.AccessControlled.AccessControlled[B])
  
  def _private[A](
    value: A
  ): morphir.ir.AccessControlled.AccessControlled[A] =
    (morphir.ir.AccessControlled.AccessControlled(
      (morphir.ir.AccessControlled.Private : morphir.ir.AccessControlled.Access),
      value
    ) : morphir.ir.AccessControlled.AccessControlled[A])
  
  def public[A](
    value: A
  ): morphir.ir.AccessControlled.AccessControlled[A] =
    (morphir.ir.AccessControlled.AccessControlled(
      (morphir.ir.AccessControlled.Public : morphir.ir.AccessControlled.Access),
      value
    ) : morphir.ir.AccessControlled.AccessControlled[A])
  
  def withAccess[A](
    access: morphir.ir.AccessControlled.Access
  )(
    ac: morphir.ir.AccessControlled.AccessControlled[A]
  ): morphir.sdk.Maybe.Maybe[A] =
    access match {
      case morphir.ir.AccessControlled.Public => 
        morphir.ir.AccessControlled.withPublicAccess(ac)
      case morphir.ir.AccessControlled.Private => 
        (morphir.sdk.Maybe.Just(morphir.ir.AccessControlled.withPrivateAccess(ac)) : morphir.sdk.Maybe.Maybe[A])
    }
  
  def withPrivateAccess[A](
    ac: morphir.ir.AccessControlled.AccessControlled[A]
  ): A =
    ac.access match {
      case morphir.ir.AccessControlled.Public => 
        ac.value
      case morphir.ir.AccessControlled.Private => 
        ac.value
    }
  
  def withPublicAccess[A](
    ac: morphir.ir.AccessControlled.AccessControlled[A]
  ): morphir.sdk.Maybe.Maybe[A] =
    ac.access match {
      case morphir.ir.AccessControlled.Public => 
        (morphir.sdk.Maybe.Just(ac.value) : morphir.sdk.Maybe.Maybe[A])
      case morphir.ir.AccessControlled.Private => 
        (morphir.sdk.Maybe.Nothing : morphir.sdk.Maybe.Maybe[A])
    }

}